### PR TITLE
Adjust to owncloud 10.2 logic

### PR DESCRIPTION
**Pull request by mrow4a**
There is a demo of the functionality:

onlyoffice-restricted-edit

Full video: 
[onlyoffice-restricted-edit.mp4.zip](https://github.com/ONLYOFFICE/onlyoffice-owncloud/files/3076323/onlyoffice-restricted-edit.mp4.zip)

@Galegod @LinneyS

### CODE REVIEW

**Comment by mrow4a** [comment]
Please also note that this prevents user from "overwritting" the file from sync client, when he does not have rights to edit content of the document and only review

**Comment by LinneyS** [comment]
But it did not work!
Opening ONLYOFFICE in the right mode is not all.
Commenting, filling in the fields and reviewing is a file change. As a result, this change should saved. But without editing rights, saving will not work. 
https://github.com/ONLYOFFICE/onlyoffice-owncloud/blob/feature/extra-permissions/controller/callbackcontroller.php#L433

**Comment by LinneyS** [comment]
This is why we need editing rights as mandatory.

**Comment by mrow4a** [comment]
@LinneyS let me test this today evening. I though you put the file as owner, thus "OnlyOffice" is privileged editor and can change the "original file", not through share. 

If this is not a case, I can change it - in principle, OnlyOffice becomes kind of a privileged mode editor and changes original file by "fileId". So you control permissions on OnlyOffice app level only..

It has to work like this for this feature, otherwise the review mode won't be respected and any user can upload a file as he pleases with sync client (sync client should only be able to get changes).

**Comment by LinneyS** [comment]
Initially, I saved the file as the owner. But later, having the author's UserId parameter, who made the changes — I redid everything. How to save a file with author, who does not have rights?

**Comment by mrow4a** [comment]
@LinneyS I am not sure if I correctly understand the problem. "Review mode" is only available with "sharing". 

With "only office review mode" which is "no edit permission in oc", and "review in onlyoffice":
- "share receiver" makes changes to the document based on what he is allowed in OnlyOffice Editor and physical document is uploaded in privileged mode by "document owner" through OnlyOffice app (using oc backend code).
- "share receiver" cannot make changes to the document outside of OnlyOffice Editor, standard oc permissions apply.

**Comment by mrow4a** [comment]
@LinneyS can you try now and QA the change?

File: controller/callbackcontroller.php [hunk 1]
```diff
@@ -373,26 +373,29 @@ public function track($doc, $users, $key, $status, $url, $token) {
...
+                    }
+
                     $userId = $users[0];
                     $user = $this->userManager->get($userId);
                     if (!empty($user)) {
-                        \OC_Util::tearDownFS();
-                        \OC_Util::setupFS($userId);
-
+                        // Login user to allow audit
                         $this->userSession->setUser($user);
```
Comment (mrow4a) [inline controller/callbackcontroller.php]: @LinneyS user must be logged in as current user

File: controller/callbackcontroller.php [hunk 2]
```diff
@@ -373,26 +373,29 @@ public function track($doc, $users, $key, $status, $url, $token) {
...
-                        if (!empty($ownerId)) {
-                            \OC_Util::setupFS($ownerId);
-                        }
                     }
 
-                    $token = isset($hashData->token) ? $hashData->token : NULL;
-                    list ($file, $error) = empty($token) ? $this->getFile($userId, $fileId) : $this->getFileByToken($fileId, $token);
+                    // Setup fs of file owner
+                    \OC_Util::tearDownFS();
+                    \OC_Util::setupFS($ownerId);
```
Comment (mrow4a) [inline controller/callbackcontroller.php]: FS however should be of file owner

File: controller/callbackcontroller.php [hunk 3]
```diff
@@ -430,6 +433,8 @@ public function track($doc, $users, $key, $status, $url, $token) {
                     }
 
                     $newData = $documentService->Request($url);
+
+                    $this->logger->debug("Track put content " . $file->getPath(), array("app" => $this->appName));
```
Comment (mrow4a) [inline controller/callbackcontroller.php]: after few seconds of closing the editor, I saw the debug msg acknowledging write to original file. Also checked on other account after restart of onlyoffice.

File: controller/callbackcontroller.php (hunk 1 above)
Comment (mrow4a) [reply controller/callbackcontroller.php]: this e.g. allows tracking in audit enterprise app who changed file.

File: controller/callbackcontroller.php (hunk 2 above)
Comment (mrow4a) [reply controller/callbackcontroller.php]: @LinneyS please remember that in OC, there are several layers. One is authentication, and other is FS access - these are completely independent, and app logic controls it.

**Comment by mrow4a** [comment]
@LinneyS other option from priviledged write is to use special core attribute e.g. 'permissions -> direct-write'. And even with "edit" permission, user wont be able to "write to file" with e.g. sync client.

@PVince81  such extra attribute plugin in OC could only go to OC 10.3 right as new feature? 10.2 is freezed ?

File: controller/callbackcontroller.php (hunk 2 above)
Comment (LinneyS) [reply controller/callbackcontroller.php]: These fixes really helped me. We will try to re-test various opening cases of the editor and check the save.

**Comment by LinneyS** [comment]
Now, without permission to edit the file from the editor is saved
And I rewrote the schema adding attributes.
233166089e315c7e0594e40dca6623c646c24680

**Comment by LinneyS** [comment]
@mrow4a How and where is the `description` property used?
https://github.com/owncloud/core/pull/33994/commits/a3e6a3e750b8a459689fd714b6537e658fe02079#r273894407

**Comment by mrow4a** [comment]
@LinneyS we decided to not use description, skip it.

**Comment by mrow4a** [comment]
@LinneyS if you wish, close this PR as your already cherry-picked commits

**Comment by LinneyS** [comment]
Thank you for help.
Expect https://github.com/owncloud/core/pull/35016

**Pull request closed without merging**